zachsinterview/src/main/java/com/imprev/zachsinterview/NumericalKeyValueStoreImpl.java

```diff
@@ -0,0 +1,111 @@
+
+    private int MAX_RETRIES = 3;
+
+    public NumericalKeyValueStoreImpl (
+            String username,
+            String secretPassword,
+            String host,
+            int port) {
+
+        this.thirdPartyAPIClient = new ThirdPartyAPIImpl(username, secretPassword, host, port);
+
+        // lets ping the partner to verify our credentials work!
+        try {
+            this.thirdPartyAPIClient.verifyCredentials();
+        } catch (IOException | TryAgainLaterException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Override
+    public void put(String key, Integer value) throws NumericalKeyValueException {
+
+        int retries = 0;
+        String errorMessage = null;
+        while (retries < MAX_RETRIES) {
+            try {
+                // we will only store one value in the list data structure the third party API uses, so if a
+                // value already exists we want to override it not append to the list.
+                List<String> values = this.thirdPartyAPIClient.get(key);
+                
+                if (values.size() == 1) {
+                    this.thirdPartyAPIClient.delete(key);
+                }
+                
+                this.thirdPartyAPIClient.add(key, String.valueOf(value));
+                
+                return;
+
+            } catch (IOException | TryAgainLaterException | InvalidCredentialsException e) {
+                errorMessage = e.getMessage();
+                retries++;
+            }
+        }
+        throw new NumericalKeyValueException(errorMessage);
+    }
+
+    @Override
+    public void delete(String key) throws NumericalKeyValueException {
+        try {
+            this.thirdPartyAPIClient.delete(key);
+            
+        } catch (IOException | TryAgainLaterException | InvalidCredentialsException e) {
+            throw new NumericalKeyValueException(e.getMessage());
+        }
+    }
+
+    @Override
+    public Integer get(String key) throws NumericalKeyValueException {
+        int retries = 0;
+        String errorMessage = null;
+        while (retries < MAX_RETRIES) {
+            try {
+                // we will only store one value in the list data structure the third party API uses
+                return Integer.parseInt(this.thirdPartyAPIClient.get(key).get(0));
+
+            } catch (Exception e) {
+                errorMessage = e.getMessage();
+                retries++;
+            }
+        }
+        throw new NumericalKeyValueException(errorMessage);
+    }
+
+    @Override
+    public Iterator<String> getKeysByPrefix(String prefix) throws NumericalKeyValueException {
+        int retries = 0;
+        String errorMessage = null;
+        while (retries < MAX_RETRIES) {
+            try {
+                return this.thirdPartyAPIClient.getKeysByPrefix(prefix);
+
+            } catch (Exception e) {
+                errorMessage = e.getMessage();
+                retries++;
+            }
+        }
```

Comment:
Improve Retry logic. The issues here are:
1. We do not need to retry on every exception, the retry behavior depends on the exception. I.e.
try {} 
catch (IOException | TryAgainLaterException e) {}
catch (InvalidCredentialsException e) {}

with the first catch, we should retry, with the second, we should not. Invalid credentials will not be fixed with retries.

2. With the retries, we shouldn't just retry immediately, we risk browning out the third party api/the situation might not be resolved if it is transient. We should retry with backoff.